table refactor: edit, delete functionalities, styling, toggling

vanessa-hu opened the pull request:
From issues #131, #132, #133

**Table Component Props:**
* columns (list of objects with each as a column; has dataField and the actual text represented (id vs 'Product ID'. also includes fields for sort (true/false), editor types (see App.tsx; like drop-down menu)
* data
* candtEditCols (set of col indices can't modify)
* canModify (boolean for if table is editable)
* modRoute: backend slug to modify this table. (ex: modRoute="/table")

**Table Component State:**
* editRows: Set of row IDs being edited
* data: from passed in prop data
* showDeleteModal (boolean, whether to pop up delete confirmation)
* rowToDelete (holds row object for one being deleted)
* columns: from props but added formatter & editorClasses field for formatting

 **Functionalities:** 
* InputFormatter:
   - formats all boxes if it's editable
* handleTryDelete & handleClickClose:
   - pops up delete confirmation modal or closes it; passed to TModal as props along with rowToDelete as the row prop
* handleEdit: 
   - adds row id of one clicked to editRows state
* handleSave: 
   - since front-end is already updated, we find the row obj inside the data with that updated info and ideally do a fetch call to update backend, then updated state to remove row.id from editRows if successful. this is commented out
* handleDelete:
   -  similar, finds + removes row obj from data & updates state. ideally would be after if a fetch call to delete is successful.
* the Fetch call
   - via one route passed into Table as modRoute prop (pass in the slash please)
   - has field "mode", with value "delete" or "edit"
   - id, with the unique row identification value
   - If it was an edit, will also have all other fields of the object, modified (including id; done via the spread operator)

 **Anything else** 
- Edit/Delete words disappear if small screen (icon only)
- people will need to modify fetched data / create columns before using Table so will have to read documentation maybe
- will need to designate one column as id

crchong1 reviewed (approved):
Looks great!

vanessa-hu commented:
Basically got all the functionalities to work. Styling unfinished. Explanations were commented under individual issues as I did them.
- empty table button/message/styling: in #136 , #138 
- pagination logic + styling: notes in #145 
- showing X of X results custom + styling: in #143 
- choosing X results per page, functionality + styling: #144
- select row state/highlight: #141 
I basically have to figure out a lot of the styling (including checkbox things / overall styles), the arrangement of bottom bar elements, probably general code cleanliness (I left a lot of console.logs everywhere). sorting wasn't included in the epic but was on the design so idk if I have to do that? I'll probably write out how the data/columns/other props need to be prepped in order to feed them into a <Table />.
- wasn't sure why codacy failed 😮

crchong1 commented on src/App.tsx:
```diff
@@ -90,11 +91,30 @@ const columns = [{
   text: 'Product Price',
   sort: true,
 }];
+const cantEdit = new Set([0]);
```
if you initialize a set, isn't it already empty?

vanessa-hu replied:
Wanted column of index 0 (first column) to be non-editable!

crchong1 commented on src/components/Base/Table.tsx:
```diff
@@ -1,192 +1,509 @@
-        this.handleSave = this.handleSave.bind(this);
-        this.deleteFormatter = this.deleteFormatter.bind(this);
+    };
+    this.handleDelete = this.handleDelete.bind(this);
```
i dont think you need to bind to this if you use arrow functions

crchong1 commented on src/components/Base/TablePageSelector.tsx:
```diff
@@ -41,6 +41,19 @@ function TablePageSelector(props: Props) : any {
   for (let i = numPagesLowerBound; i <= numPagesUpperBound; i += 1) {
     numPagesArray.push(i + 1);
   }
+  let pageDisplayNums : number[];
```
can we convert some of these numbers to enums? there seems to be some arbitrary numbers (6, 3, 4, etc) so enums can help readers understand what these values actually represent

vanessa-hu replied:
I wasn't completely sure what some of that code did, but I can make some constants / try to see what they mean!

crchong1 commented on src/components/Base/TablePageSelector.tsx:
```diff
@@ -58,11 +71,21 @@ function TablePageSelector(props: Props) : any {
-          {currentPage < (numPages - 1) ? <li className="page-item"><span className="page-link" onClick={handleClickNext}>&raquo;</span></li> : <div />}
+      <nav aria-label="Page navigation">
+        <ul className="pagination">
+          {(currentPage > 0 && numPages > 5) ? <li className="page-item"><span className="page-link" onClick={handleClickPrevious}>&laquo;</span></li> : <div style={{ width: '2.2em' }} />}
```
I would remove contants and move them to static final consts at the top

crchong1 commented on src/static/styles/App.scss:
```diff
@@ -1,624 +1,764 @@
+    margin: auto;
+}
 
-// class for rows being edted
```
can you move all table scss into a different scss file (and then import it into app) before things get too confusing?

vanessa-hu replied:
Yes! Based on [this](https://stackoverflow.com/questions/17598996/sass-use-variables-across-multiple-files) and [this](https://sass-lang.com/documentation/at-rules/use) and [this](https://github.com/material-components/material-components-web/issues/6431), @import will eventually be deprecated -> should use @use but we don't have Dart Sass so that doesn't work yet.

ideally we’d use @use ‘App’ and access variables like App.$color-name; otherwise everything is global / a bit confusing

crchong1 commented on src/static/styles/App.scss:
```diff
@@ -1,624 +1,764 @@
+    font-size: 1.2rem;
+}
+
+// .react-bootstrap-table {
```
I would remove these lines that are commented out

crchong1 reviewed (changes requested):
Looks really good- just added some comments. Most of these comments are around readability, as I haven't gotten the chance to look at the functionality yet

vanessa-hu commented:
- remember: when I change pages without hitting save, the row is still highlighted in edit mode when i go back to that page, but resets the edit button. should prob fix that

loafyyy commented:
Good job @vanessa-hu! Overall, the table looks quite good. 

One glitch I noticed while testing it is that sorting the table while editing one of the rows can lead to situations like below where the Edit column shows "Edit" instead of "Save" when you're already editing it:
![image](https://user-images.githubusercontent.com/25329899/102395206-d07f6e80-3fa8-11eb-9d39-b7e79b795763.png)
To reproduce the glitch you can click Edit on the first row, sort the Product Price twice, then click Edit on the first row again, then sort the Product Price again. I think a quick fix would just to prevent sorting on edit if that's possible.

Also, I'm just wondering, what is the point of the checkboxes? Because clicking Edit, Save, or Delete only performs the action on the given row, not all the selected/checkboxed rows.

vanessa-hu commented:
I don't think you can prevent sorting on edit since whether a column can be sorted, is passed in with the columns prop. But, there is an onSort listener with bootstrap table 2, so I can clear selectRows and editRows whenever a table is resorted ([link here](https://react-bootstrap-table.github.io/react-bootstrap-table2/docs/basic-sort.html)).

I'm not sure exactly why the select option is there, but it was in the figma design! based on the epic explanation #142 it seems like it was a good-to-have in case we want to do something specific in a parent component, so it's just stored in a state for now? @loafyyy

vanessa-hu commented:
Combined with changes in server pull request [here](https://github.com/keepid/keepid_server/pull/76), the adminPanel can be displayed via the new Table component! Search bar is still specific to adminPanel, but also works still.
Fixed some bugs in pagination and added canSelect prop to denote whether we want a select column on the table.

For future tables that need edit/delete, that functionality will depend on the onEdit and onDelete props passed into the Table. those should make sure to communicate with the backend to make changes, then pass the updated data into Table as a prop so everything is actually updated. Currently the dummy table-test "works" because there's a state var in Table called data (that's a copy of the prop) that I modify, to try out what delete would look like (visually). but that could be removed in the future because of the passed-in prop updating via onEdit and onDelete that should auto-update the data to have different row info/deleted rows.


https://user-images.githubusercontent.com/21111373/103573506-5b62d300-4e9c-11eb-9007-f59859f1bc2a.mp4
